integration-webhook/src/main/java/org/symphonyoss/integration/webhook/parser/metadata/MetadataParser.java

```diff
@@ -0,0 +1,279 @@
+
+    preProcessInputData(node);
+    processMetadataObjects(root, node, objects);
+    postProcessOutputData(root, node);
+
+    try {
+      Map<String, Object> result = new LinkedHashMap<>();
+      result.put(metadata.getName(), root);
+
+      return JsonUtils.writeValueAsString(result);
+    } catch (JsonProcessingException e) {
+      LOGGER.error("Fail to parse incoming payload", e);
+      return null;
+    }
+  }
+
+  /**
+   * Perform a pre-processing on the input data.
+   *
+   * This method must be implemented by the concrete parser classes to perform any data manipulation
+   * on the JSON input payload or augment the payload received with internal Symphony data retrieved
+   * through the Symphony API's.
+   *
+   * Examples:
+   * - Replace '\n' to '<br/>' tags
+   * - Uppercase JSON field content
+   * - Scape characters
+   * - Include user ID and username retrieved from the User API
+   *
+   * @param input JSON input payload
+   */
+  protected abstract void preProcessInputData(JsonNode input);
+
+  /**
+   * Process metadata objects to generate Entity JSON. This method is called recursively for
+   * the nested metadata objects. The metadata fields inside each metadata object have your own
+   * logic to build the corresponding field into Entity JSON.
+   *
+   * Basically, this method gets the JSON object received from the third-party service and generates
+   * the Entity JSON intended according to the metadata objects defined in a XML document.
+   *
+   * Each metadata object defined must generate a JSON object inside the Entity JSON and the metadata
+   * fields must generate fields inside those JSON objects.
+   *
+   * @param root Root object from Entity JSON
+   * @param node JSON node received from the third-party service
+   * @param objects List of metadata objects
+   */
+  private void processMetadataObjects(EntityObject root, JsonNode node, List<MetadataObject> objects) {
+    for (MetadataObject object : objects) {
+      EntityObject entity = new EntityObject(object.getType(), object.getVersion());
+
+      if (object.getFields() != null) {
+        for (MetadataField field : object.getFields()) {
+          field.process(entity, node);
+        }
+      }
+
+      if (object.getChildren() != null) {
+        processMetadataObjects(entity, node, object.getChildren());
+      }
+
+      if (!entity.getContent().isEmpty()) {
+        root.addContent(object.getId(), entity);
+      }
+    }
+  }
+
+  /**
+   * Perform a post-processing on the output data.
+   *
+   * This method must be implemented by the concrete parser classes to include additional fields
+   * on the output Entity JSON which weren't be processed directly from the JSON input payload.
+   *
+   * Example:
+   * - Array of JSON objects
+   *
+   * @param output Output Entity JSON
+   * @param input JSON input payload
+   */
+  protected abstract void postProcessOutputData(EntityObject output, JsonNode input);
```

Comment:
This doesn't need to be abstract. You can provide an implementation that does nothing. That way if subclass doesn't need to do any post-processing, it doesn't need to implement a do-nothing method.